### PR TITLE
chore(broker): adds WorkflowInstanceCreationRecord

### LINKED ISSUE #2087: Introduce new record to create workflow instances
In order to remove the payload field from the `WorkflowInstanceRecord` (#1852), we first need to migrate creation of workflow instances to a different record type. The cleanest and simplest solution is to add a new record type, creating a workflow instance and setting its variables must be done atomically.

Related to #1852

### LINKED ISSUE #2092: Add processor for WorkflowInstanceCreationRecord#CREATE
When processing `WorkflowInstanceCreationRecord#CREATE`, we should:

1. reject the command if the workflow definition does not exist (whether specified by bpmnProcessId and version, or key)
1. reject the command if the definition has no none start event
1. reject the command if the variables is invalid
1. publish a new `ELEMENT_ACTIVATING` event for the workflow instance to kickstart it
1. set the variables for the new workflow instance (publishing `Variable#CREATED` records for each)
1. publish a `WorkflowInstanceCreationRecord#CREATED` follow up event as acknowledgement

Related to #2087

### PR DESCRIPTION
**Pull request by npepinpe**
- adds new WorkflowInstanceCreationRecord with two intents, CREATE and CREATED
- adds processor for CREATE intent
- implements equals/hashCode for msgpack values to ease comparing `UnpackedObject`

Closes #2087
Closes #2092

### CODE REVIEW

**Comment by npepinpe** [comment]
This does not remove code that it will eventually replace, as this will be done in one of the follow up issues (see issues related to #1852, which is the primary motivation for this)

File: broker-core/src/main/java/io/zeebe/broker/workflow/processor/instance/CreateWorkflowInstanceProcessor.java [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    final DeployedWorkflow workflow;
+
+    if (record.getKey() >= 0) {
+      workflow = getWorkflow(record.getKey(), controller);
+    } else {
+      final DirectBuffer bpmnProcessId = record.getBpmnProcessId();
+
+      if (bpmnProcessId.capacity() == 0) {
+        controller.reject(RejectionType.INVALID_ARGUMENT, ERROR_MESSAGE_NO_IDENTIFIER_SPECIFIED);
+        workflow = null;
```
Comment (MiguelPires) [inline broker-core/src/main/java/io/zeebe/broker/workflow/processor/instance/CreateWorkflowInstanceProcessor.java]: Why null workflow here and not on declaration?

**Review by miguelpires — CHANGES_REQUESTED** [review]
Looks good. Just pointed out two things

File: broker-core/src/main/java/io/zeebe/broker/workflow/processor/instance/CreateWorkflowInstanceProcessor.java (hunk 1 above)
Comment (npepinpe) [reply broker-core/src/main/java/io/zeebe/broker/workflow/processor/instance/CreateWorkflowInstanceProcessor.java]: So I can have a final variable (i.e. safer, compiler enforced single assignment). Having a single place to assign a return value, and a single return statement, just usually means its easier to think about possible cases (e.g. cyclomatic complexity)

File: broker-core/src/main/java/io/zeebe/broker/workflow/processor/instance/CreateWorkflowInstanceProcessor.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    elementInstanceState.flushDirtyState();
+
+    return instance;
+  }
+
+  private DeployedWorkflow getWorkflow(
+      WorkflowInstanceCreationRecord record, CommandControl controller) {
+    final DeployedWorkflow workflow;
+
+    if (record.getKey() >= 0) {
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply broker-core/src/main/java/io/zeebe/broker/workflow/processor/instance/CreateWorkflowInstanceProcessor.java]: This is semantics I guess :sweat_smile: 

Technically our generator always returns a number from a sequence `[1, +inf[`, but we also say that `-1` is our null value here, so `0` is kind of a...meaningless number in this case. I just felt it's simpler to go with the second part, e.g. it's `> -1`.

[the remark this answers is not in the document]
Comment (Zelldon) [reply broker-core/src/main/java/io/zeebe/broker/workflow/processor/instance/CreateWorkflowInstanceProcessor.java]: I think this is not true I saw cases (especially tests where we use zero as key)

[the remark this answers is not in the document]
Comment (npepinpe) [reply broker-core/src/main/java/io/zeebe/broker/workflow/processor/instance/CreateWorkflowInstanceProcessor.java]: That's also true :sweat_smile: 

But our generate starts at 0 and we always call `nextValue` which increments and return, so in practice I think it's always 1? I would say that fact is "not part of the spec" in this case, so better use `>= 0`?

[the remark this answers is not in the document]
Comment (MiguelPires) [reply broker-core/src/main/java/io/zeebe/broker/workflow/processor/instance/CreateWorkflowInstanceProcessor.java]: Yeah, that's why I mentioned it (the generator starts at 0 and nextValue adds 1) and the "old" creation handler gets the workflow by key if ``` key > 0```. If zero is undefined shouldn't the processor fall back to using the process id? Anyway, it's not super relevant so I'll leave it up to you

**Review by miguelpires — APPROVED** [review]

**Comment by menski** [comment]
bors retry

**Pull request merged**